tokio/src/sync/once_cell.rs

```diff
@@ -0,0 +1,334 @@
+use super::Semaphore;
+use crate::loom::cell::UnsafeCell;
+use std::error::Error;
+use std::fmt;
+use std::future::Future;
+use std::mem::MaybeUninit;
+use std::ptr;
+use std::sync::atomic::{AtomicBool, Ordering};
+
+/// A thread-safe cell which can be written to only once.
+///
+/// Provides the functionality to either set the value, in case `OnceCell`
+/// is uninitialized, or get the already initialized value by using an async
+/// function via [`OnceCell::get_or_init`].
+///
+/// [`OnceCell::get_or_init`]: crate::sync::OnceCell::get_or_init
+///
+/// # Examples
+/// ```
+/// use tokio::sync::OnceCell;
+///
+/// async fn some_computation() -> u32 {
+///     1 + 1
+/// }
+///
+/// static ONCE: OnceCell<u32> = OnceCell::const_new();
+///
+/// #[tokio::main]
+/// async fn main() {
+///     let result1 = tokio::spawn(async {
+///         ONCE.get_or_init(some_computation).await
+///     }).await.unwrap();
+///     assert_eq!(*result1, 2);
+/// }
+/// ```
+pub struct OnceCell<T> {
+    value_set: AtomicBool,
+    value: UnsafeCell<MaybeUninit<T>>,
+    semaphore: Semaphore,
+}
+
+impl<T> Default for OnceCell<T> {
+    fn default() -> OnceCell<T> {
+        OnceCell::new()
+    }
+}
+
+impl<T: fmt::Debug> fmt::Debug for OnceCell<T> {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt.debug_struct("OnceCell")
+            .field("value", &self.get())
+            .finish()
+    }
+}
+
+impl<T: Clone> Clone for OnceCell<T> {
+    fn clone(&self) -> OnceCell<T> {
+        OnceCell::new_with(self.get().map(|v| (*v).clone()))
+    }
+}
+
+impl<T: PartialEq> PartialEq for OnceCell<T> {
+    fn eq(&self, other: &OnceCell<T>) -> bool {
+        self.get() == other.get()
+    }
+}
+
+impl<T: Eq> Eq for OnceCell<T> {}
+
+impl<T> OnceCell<T> {
+    /// Creates a new uninitialized OnceCell instance.
+    pub fn new() -> Self {
+        OnceCell {
+            value_set: AtomicBool::new(false),
+            value: UnsafeCell::new(MaybeUninit::uninit()),
+            semaphore: Semaphore::new(1),
+        }
+    }
+
+    /// Creates a new initialized OnceCell instance if `value` is `Some`, otherwise
+    /// has the same functionality as [`OnceCell::new`].
+    ///
+    /// [`OnceCell::new`]: crate::sync::OnceCell::new
+    pub fn new_with(value: Option<T>) -> Self {
+        let (value_set, value) = if let Some(v) = value {
+            (AtomicBool::new(true), UnsafeCell::new(MaybeUninit::new(v)))
+        } else {
+            (
+                AtomicBool::new(false),
+                UnsafeCell::new(MaybeUninit::uninit()),
+            )
+        };
+        OnceCell {
+            value_set,
+            value,
+            semaphore: Semaphore::new(1),
+        }
```

Comment:
You need to close the semaphore in the case where the value is set.
```suggestion
        if let Some(v) = value {
            let semaphore = Semaphore::new(0);
            semaphore.close();
            OnceCell {
                value_set: AtomicBool::new(true),
                value: UnsafeCell::new(MaybeUninit::new(v)),
                semaphore,
            }
        } else {
            OnceCell::new()
        }
```

Reply:
I agree that this is more consistent, but it should't matter in practice since we can only acquire the permit when `value_set` is true, which we set to false if value is `Some`.

Reply:
I do want the semaphore to be closed here, because it makes the safety arguments you have elsewhere easier to follow if the semaphore is always closed when the value is set.